packages/paid-storage-contract/src/contract.ts

```diff
@@ -185,4 +211,287 @@ class UserStorage implements StorageManagement {
+    }
+    return this.filemap.get(file)
+  }
+
+  /**
+   * Get a specific file partition
+   * @param account_id owner of the file
+   * @param filename name of the file
+   * @param part partition index
+   */
+  @view({})
+  get_partition({ application, account_id, filename, part }: { application: string, account_id: string, filename: string, part: number }): string {
+    const files = this.applications.get(`${account_id}/${application}`)?.currentFiles || [];
+    // @ts-ignore
+    const filepath = files.find((f: string) => f.endsWith(filename));
+    return this.partitions.get(`${filepath}-${part}`)
+  }
+
+  /**
+   * Begin application deploy
+   * @param application app to which the file belongs
+   */
+  @call({})
+  deploy_application({ application, files }: { application: string, files: string[] }) {
+    const startingSize = near.storageUsage();
+    const appKey = this.buildApplicationKey(application, near.signerAccountId());
+    let app = this.applications.get(appKey);
+    if (!app) {
+      app = new Application();
+      this.applications.set(appKey, app);
+    }
+    this.startDeployment({ app, account_id: near.signerAccountId(), application, files });
+    const diff = near.storageUsage() - startingSize;
+    this.pay_storage_cost(diff);
+    return { diff };
+  }
+
+  /**
+   * Begin application deploy
+   * @param application app to which the file belongs
+   */
+  @call({})
+  post_deploy_cleanup({ application }: { application: string }): { files: string[], diff: bigint } {
+    const startingSize = near.storageUsage();
+    const app = this.applications.get(`${near.signerAccountId()}/${application}`)!;
+    const files = app.currentFiles;
+    this.completeDeployment({ app, account_id: near.signerAccountId(), application });
+    const diff = near.storageUsage() - startingSize;
+    this.pay_storage_cost(diff);
+    return { files, diff };
+  }
+
+  @call({})
+  delete_application({ application }: { application: string }) {
+    const startingSize = near.storageUsage();
+    this.applications.remove(`${near.predecessorAccountId()}/${application}`)
+    const diff = near.storageUsage() - startingSize;
+    this.pay_storage_cost(diff);
+    return { diff };
+  }
+
+  /**
+   * Upload a file partition
+   * @param application app to which the file belongs
+   * @param filename name of the file
+   * @param bytes binary file partition data encoded as base64 substring
+   * @param part partition index
+   * @param totalParts total number of partitions for this file
+   */
+  @call({})
+  upload_file_partition({ application, filename, bytes, part, totalParts }: { application: string, filename: string, bytes: string, part: number, totalParts: number }) {
+    const startingSize = near.storageUsage();
+    let ret = { filemapDiff: 0n, partitionsDiff: 0n };
+    let diff = -1n;
+    let error: string | null = null;
+    try {
+      const app = this.applications.get(`${near.signerAccountId()}/${application}`);
```

Comment:
my understanding is that it is best practice to use predecessor instead of signer unless you have a specific flow that requires signer e.g. a marketplace